pkg/session/syssession/pool.go

```diff
@@ -0,0 +1,259 @@
+	WithSession(func(*Session) error) error
+}
+
+// AdvancedSessionPool is a recyclable resource pool for the system internal session.
+type AdvancedSessionPool struct {
+	noopOwnerHook
+	ctx     context.Context
+	pool    chan *session
+	factory Factory
+	mu      struct {
+		sync.RWMutex
+		closed bool
+		cancel context.CancelFunc
+	}
+}
+
+// NewAdvancedSessionPool creates a default session pool with the given capacity and factory function.
+func NewAdvancedSessionPool(capacity int, factory Factory) *AdvancedSessionPool {
+	intest.AssertNotNil(factory)
+	if capacity <= 0 || capacity > PoolMaxSize {
+		intest.Assert(suppressAssertInTest, "invalid capacity: %d", capacity)
+		capacity = PoolMaxSize
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	ctx = kv.WithInternalSourceType(ctx, kv.InternalTxnOthers)
+	pool := &AdvancedSessionPool{
+		ctx:     ctx,
+		pool:    make(chan *session, capacity),
+		factory: factory,
+	}
+	pool.mu.cancel = cancel
+	return pool
+}
+
+func (p *AdvancedSessionPool) getInternal() (s *session, _ error) {
+	select {
+	case r, ok := <-p.pool:
+		if !ok {
+			return nil, errors.New("session pool closed")
+		}
+		return r, nil
+	default:
+		// the pool is empty, continue to create a new session
+	}
+
+	sctx, err := p.factory()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	defer func() {
+		if s == nil {
+			// s == nil means the internal session is not successfully created.
+			// close it instead.
+			sctx.Close()
+		}
+	}()
+
+	return newInternalSession(sctx, p)
+}
+
+// Get gets a session from the session pool.
+func (p *AdvancedSessionPool) Get() (*Session, error) {
+	internal, err := p.getInternal()
+	if err != nil {
+		return nil, err
+	}
+
+	intest.AssertNotNil(internal)
+	intest.Assert(!internal.IsClosed())
+	intest.Assert(internal.Owner() == p)
+	se := &Session{}
+	defer func() {
+		if se.internal != internal {
+			// se.internal != internal means the session is not successfully created.
+			// We need to close the internal session
+			internal.Close()
+		}
+	}()
+
+	if err = internal.TransferOwner(p, se); err != nil {
+		return nil, err
+	}
+
+	se.internal = internal
+	return se, nil
+}
+
+// Put puts the session back to the pool.
+func (p *AdvancedSessionPool) Put(se *Session) {
+	if se == nil || se.internal == nil || se.internal.Owner() != se {
+		// Do nothing when the input session is invalid.
+		return
+	}
+
+	internal := se.internal
+	// We should transfer the owner back to the pool first for reasons:
+	// 1. Make sure the input Session is valid by checking the internal session's owner.
+	// 2. After ownership is transferred back to pool, we can ensure only the pool can access the internal session.
+	if err := internal.TransferOwner(se, p); err != nil {
+		// Notice that we should not call `internal.Close` to make sure only close the internal session when its owner
+		// is the current session.
+		logutil.BgLogger().Warn(
```

Comment:
Could you give an example here?

Reply:
I've added more comments:
https://github.com/pingcap/tidb/pull/60168/commits/8e7e51917deb120e4bdc6edd12c5ad1b49ada34f